### PR TITLE
document meta.yml

### LINKED ISSUE #19: documentation on meta.yml
It would be nice if there was a template for the meta.yml (or a link to a template).

### PR DESCRIPTION
**Pull request by liyishuai**
Resolve #19

### CODE REVIEW

File: ref.yml [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+- circleci:
+    type: bool
+    used:
+    - README.md
+- community:
+    type: bool
+    used:
+    - README.md
+    - index.md
+- coqdoc:
```
Comment (Zimmi48) [inline ref.yml:42]: I hadn't realized we had this boolean variable available already. We should generate `index.md` only if it is set to `true`, right?

File: ref.yml [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    - README.md
+- license:
+    required: false
+    keys:
+    - fullname:
+        required: true
+    - identifier:
+        required: true
+    - file:
+        required: false
```
Comment (Zimmi48) [inline ref.yml:91]: I suggest adding the default value:
```suggestion
        required: false
        default: LICENSE
```

File: ref.yml [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    keys:
+    - text:
+        required: true
+    - coq.opam:
+        required: true
+    used:
+    - README.md
+    - coq.opam
+- plugin:
+    used:
```
Comment (Zimmi48) [inline ref.yml:119]:
```suggestion
    type: bool
    used:
```

File: ref.yml [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+- doi:
+    required: false
+    used:
+    - README.md
+- synopsis:
+    required: true
+    used:
+    - coq.opam
+- description:
+    requied: true
```
Comment (Zimmi48) [inline ref.yml:56]:
```suggestion
    required: true
```

File: ref.yml [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    fields:
+    - description:
+        required: true
+    - opam:
+        required: false
+        keys:
+        - name:
+            required: true
+        - version
+    - nix
```
Comment (Zimmi48) [inline ref.yml:131]:
```suggestion
    - nix
        required: true
```

File: ref.yml [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    - extracted_build:
+        required: false
+    - extracted_make_target:
+        required: true
+    used:
+    - README.md
+    - extracted.opam
+    - .travis.yml
+- documentation:
+    used:
```
Comment (Zimmi48) [inline ref.yml:207]:
```suggestion
    required: false
    used:
```

File: ref.yml [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    - README.md
+- opam-file-maintainer:
+    required: false
+    used:
```
Comment (Zimmi48) [inline ref.yml:211]:
```suggestion
    default: palmskog@gmail.com
    used:
```

**Comment by liyishuai** [comment]
> Should we use additional YAML syntax or YAML comments to explain them?

We can add `description` to the keys and fields.

> Is the distinction between `fields` for lists and `keys` for sets something standard? The naming is not intuitive to me.

Not standard, better namings solicited.

> It would be great to have a CI check that the mustache templates respect this reference, but it does not look easy to do, unless tools already exist for this?

I'm unaware of existing tools that can check mustache templates, but writing this document has forced me to review all the templates and revealed some problems:
- https://github.com/coq-community/templates/blob/cb4cced445482f1c8286f5beeb615ef0b105d59f/.travis.yml.mustache#L60-L69 There should be a separate `opam-file-version` for extracted package. 
- https://github.com/coq-community/templates/blob/cb4cced445482f1c8286f5beeb615ef0b105d59f/README.md.mustache#L66-L70 When `community=true` and `maintainers` is empty, this produces
    ```markdown
    - Coq-community maintainer(s): 
    ```
    without any maintainer, which seems awkward.

To check mustache templates, the current `required` is too coarse-grained. For example, `supported_coq_versions` and `supported_ocaml_versions` have two subkeys: `text` is required by README, while `opam` is required by OPAM files. Missing `opam` does not affect README.

**Comment by Zimmi48** [comment]
> > Is the distinction between `fields` for lists and `keys` for sets something standard? The naming is not intuitive to me.
> 
> Not standard, better namings solicited.

What about "fields" and "item_fields"?

We could document that `maintainers` is required if `community` is true.

> To check ~mustache templates~`meta.yml`, the current `required` is too coarse-grained. For example, `supported_coq_versions` and `supported_ocaml_versions` have two subkeys: `text` is required by README, while `opam` is required by OPAM files. Missing `opam` does not affect README.

Yep, I saw that too. Would it work to move the `used` fields to the subkeys in such cases?

>What property of mustache templates do we need to check? Whether it depends on any not-specified-as-required key?

We could check many things:
- are there undocumented keys?
- are there keys that are not documented as required but should?
- are there any omissions in `used` or some files listed which could be removed?

Anyway, I'm fine with this PR in the current state!

**Comment by erikmd** [comment]
Hi @liyishuai @Zimmi48 , thanks for adding this human-friendly documentation.

BTW shouldn't the `ref.yml` file mentioned from the `README.md` (e.g., with an hyperlink)?

**Comment by Zimmi48** [comment]
> BTW shouldn't the `ref.yml` file mentioned from the `README.md` (e.g., with an hyperlink)?

Yes! This was also something I requested but @liyishuai probably missed it.

**Comment by liyishuai** [comment]
I'm thinking of rendering a more readable webpage from this file and mention that user-friendly version in README. Is there a tool for that or should we write another mustache?

**Comment by Zimmi48** [comment]
Ah OK, I reviewed this as if it would be the actual, user-friendly, documentation. To me, it looks reasonably accessible. And in any case, since it exists now, it should be mentioned in the README until there is something better to show.

**Pull request merged**